readme.md
Gesis
=====

[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/gesis)](http://cran.r-project.org/package=gesis) [![](http://cranlogs.r-pkg.org/badges/grand-total/gesis)](http://cran.r-project.org/web/packages/gesis)

Introduction
------------

The [GESIS Data Catalogue](https://dbk.gesis.org/) offers a repository of approximately 5,000 datasets.

To install the package from github:

``` r
# install.packages("devtools")
devtools::install_github("expersso/gesis")
```

``` r
library(gesis)
```

A simple example
----------------

We start by listing all available groups of studies:

``` r
groups <- get_study_groups()
head(groups, 10)
```

    ##    group_no                                        value
    ## 1      0001 International Social Survey Programme (ISSP)
    ## 2      0002                     EB - Flash Eurobarometer
    ## 3      0003                               Travel Surveys
    ## 4      0004                            Time Budget Study
    ## 5      0005       EB - Central and Eastern Eurobarometer
    ## 6      0006       EB - Candidate Countries Eurobarometer
    ## 7      0007                                       ALLBUS
    ## 8      0008      EB - Standard and Special Eurobarometer
    ## 9      0009                  European Values Study (EVS)
    ## 10     0010                               Politbarometer

We see that the Eurobarometer has study group number 0008 Let's looks at all available Eurobarometer waves:

``` r
eurobars <- get_datasets("0008")
head(eurobars)
```

    ##    doi                           title
    ## 1 0078 Attitudes towards Europe (1962)
    ## 2 0626 European Communities Study 1970
    ## 3 0627 European Communities Study 1971
    ## 4 0628 European Communities Study 1973
    ## 5 0986  Eurobarometer 2 (Oct-Nov 1974)
    ## 6 0987      Eurobarometer 3 (May 1975)

We would now like to download the first three studies. We first need to log in to the Gesis website and then pass the DOIs (unique data set identifiers) to `download_dataset`:

``` r
# username and password stored as environment 
# variables "GESIS_USER" and "GESIS_PASS"
gesis_session <- login()
```

``` r
if(!dir.exists("downloads")) dir.create("downloads")
download_dataset(s = gesis_session, doi = eurobars$doi[1:3], 
                 path = "downloads", filetype = ".dta")
```

    ## Downloading DOI: 0078

    ## Downloading DOI: 0626

    ## Downloading DOI: 0627

``` r
(files <- list.files("downloads", full.names = TRUE))
```

    ## [1] "downloads/ZA0078_v1-0-1.dta" "downloads/ZA0626_v1-0-1.dta"
    ## [3] "downloads/ZA0627_v1-0-1.dta"

We can also download the codebooks for the same studies:

``` r
download_codebook(eurobars$doi[1:3], path = "downloads")
```

    ## Downloading codebook for DOI: 0078

    ## Downloading codebook for DOI: 0626

    ## Downloading codebook for DOI: 0627

Using the `haven` package we can now read the data sets:

``` r
library(haven)
df <- read_dta(files[1])
dim(df)
```

    ## [1] 4774  175

Disclaimer: the `gesis` package is neither affiliated with, nor endorsed by, the Leibniz Institute for the Social Sciences. I have been unable to find any indication that programmatic access to the website is disallowed under its terms of use (indeed, its [guidelines](https://dbk.gesis.org/dbksearch/guidelines.asp) appear to encourage it). That said, I would discourage users from using the `gesis` package to put undue pressure on their servers by initiating unnecessary (or unnecessarily large) batch downloads.


NEWS.md
# gesis 0.2

* Completely redesigned package to remove dependency on RSelenium


tests/testthat.R
library(testthat)
library(gesis)

test_check("gesis")


tests/testthat/test-availability.R
context("Available datasets")

test_that("get_study_groups() works", {
    groups <- get_study_groups()
    expect_equal(class(groups), c("tbl_df", "tbl", "data.frame"))
    expect_gt(nrow(groups), 0)
})

test_that("get_datasets() works", {
    dfs <- get_datasets("0001")
    expect_equal(class(dfs), c("tbl_df", "tbl", "data.frame"))
    expect_gt(nrow(dfs), 0)
})


tests/testthat/test-main_functionality.R
context("Main functionality")

if(!identical(Sys.getenv("GESIS_USER"), "")) {
    test_that("login works", {
      s <- login()
      expect_equal(class(s), "session")
    })
}


R/gesis.R
#' Log in to the Gesis website
#'
#' @param username Your Gesis username
#' @param password Your Gesis password
#'
#' @details
#' The username and password can also be stored as enviroenment variables
#' "GESIS_USER" and "GESIS_PASS" so as not to store these in plaintext in a
#' script.
#'
#' @return A session object
#' @export
#' @import rvest
#' @import xml2
#'
#' @examples
#' \dontrun{s <- login("my_gesis_username", "my_gesis_password")}
login <- function(username = "", password = "") {
    if(username == "") username <- Sys.getenv("GESIS_USER")
    if(password == "") password <- Sys.getenv("GESIS_PASS")

    if(any(username == "", password == "")) {
        stop("Please provide username and/or password.", call. = FALSE)
    }

    url <- "https://dbk.gesis.org/dbksearch/index.asp"
    s <- html_session(url)
    form <- html_form(s)[[1]]
    form <- set_values(form, user = username, pass = password)
    form$url <- ""
    suppressMessages(submit_form(s, form))
}

#' Download a Gesis data set
#'
#' @param s A session object created with login()
#' @param doi The unique identifier(s) for the data set(s)
#' @param path Directory to which to download the file
#' @param filetype The filetype to download (usually available: .dta/.por/.sav)
#' @param purpose The purpose for downloading the data. See details.
#' @param quiet Whether to output download message.
#'
#' @details Datasets reposited with GESIS are uniquely identified with a
#'   numberic identifier called a "DOI". This identifier appears both in the URL
#'   for a dataset's website, and on the website itself.
#'
#'   In addition to accepting the terms of use, you need to input a purpose for
#'   downloading a data set. The options are as follows:
#'
#' 1. for scientific research (incl. PhD)
#' 2. for reserach with commercial mandate
#' 3. for teaching as lecturer
#' 4. for my academic studies
#' 5. for my final exam (e.g. bachelor or master)
#' 6. for professional training and qualification
#'
#' @return Nothing
#' @export
#' @import rvest
#' @import xml2
#' @import httr
#'
#' @examples
#' \dontrun{s <- login("my_gesis_username", "my_gesis_password")
#' download_dataset(s, doi = "0078")}
download_dataset <- function(s, doi, path = ".", filetype = ".dta",
                             purpose = 1, quiet = FALSE) {
    for(d in doi) {

        url <- paste0("https://dbk.gesis.org/dbksearch/SDesc2.asp?db=E&no=", d)
        s <- jump_to(s, url)
        stop_for_status(s)

        s <- suppressMessages(
            follow_link(s, xpath = sprintf("//a[contains(text(), '%s')]", filetype))
        )
        stop_for_status(s)

        form <- html_form(s)[[2]]
        form <- set_values(form, zweck = 1, projectok = 1)
        form$url <- ""
        s <- suppressMessages(submit_form(s, form))
        stop_for_status(s)

        if(!quiet) message("Downloading DOI: ", d)
        filename <- gsub("^.*?\"|\"", "", s$response$headers$`content-disposition`)
        filename <- file.path(path, filename)
        writeBin(content(s$response, "raw"), filename)
    }
}

#' Download the codebook for a Gesis data set
#'
#' @param doi The unique identifier(s) for the data set(s)
#' @param path Directory to which to download the file
#' @param quiet Whether to output download message.
#'
#' @return Nothing
#' @export
#' @import rvest
#' @import xml2
#' @import httr
#'
#' @examples
#' download_codebook(doi = "0078")
download_codebook <- function(doi, path = ".", quiet = FALSE) {
    for(d in doi) {

        url <- paste0("https://dbk.gesis.org/dbksearch/SDesc2.asp?db=E&no=", d)
        page <- read_html(url)
        node <- html_nodes(page, xpath = "//a[contains(text(), '_cdb')]")
        node <- paste0("https://dbk.gesis.org/dbksearch/", html_attr(node, "href"))
        resp <- GET(node)

        if(!quiet) message("Downloading codebook for DOI: ", d)
        filename <- gsub("^.*?\"|\"", "", resp$headers$`content-disposition`)
        filename <- file.path(path, filename)
        writeBin(content(resp, "raw"), filename)
    }
}

#' Get a dataframe with all available groups of studies
#'
#' @return A dataframe
#' @export
#'
#' @import rvest
#' @import xml2
#'
#' @examples
#' groups <- get_study_groups()
#' head(groups)
get_study_groups <- function() {
    url <- "https://dbk.gesis.org/dbksearch/gdesc.asp"
    page <- read_html(url)

    group_no <- html_attr(html_nodes(page, xpath = "//input//parent::a//input"), "name")
    group_no <- gsub("TI", "", group_no)
    value <- html_attr(html_nodes(page, xpath = "//input//parent::a//input"), "value")

    df <- data.frame(group_no, value, stringsAsFactors = FALSE)
    class(df) <- c("tbl_df", "tbl", "data.frame")
    df
}

#' Get a dataframe of all individual data sets within a group of studies
#'
#' @param group_no The group number (usually obtained from get_study_groups())
#'
#' @return A dataframe
#' @export
#'
#' @import rvest
#' @import xml2
#'
#' @examples
#' # Get DOIs and titles for all Eurobarometer studies
#' eurobars <- get_datasets("0008")
#' head(eurobars)
get_datasets <- function(group_no) {
    url <- paste0("https://dbk.gesis.org/dbksearch/GDESC2.asp?db=e&no=", group_no)
    page <- read_html(url)

    nodes <- html_nodes(page, xpath = "//li//a[contains(@href, 'no=')]")
    text <- html_nodes(page, xpath = "//li//a[contains(@href, 'no=')]//parent::li")
    text <- html_text(text, TRUE)

    doi <- substr(text, 3, 6)
    title <- substr(text, 8, stop = 10000L)

    df <- data.frame(doi, title, stringsAsFactors = FALSE)
    class(df) <- c("tbl_df", "tbl", "data.frame")
    df
}
